src/lib/openssl_crypto.c

```diff
@@ -861,165 +867,138 @@ openssl_read_pem_seckey(const char *f, pgp_key_t *key, const char *type, int ver
+      botan_pk_op_encrypt(op_ctx, rng, bt_ciphertext, &out_len, in, length)) {
+
+      FAIL("Encryption fails");
+  }
+
+  /*
+   * Botan's ElGamal formats the g^k and msg*(y^k) together into a single byte string.
+   * We have to parse out the two values after encryption, as rnp stores those values separatelly.
+   */
+  memcpy(g2k, bt_ciphertext, p_len);
+  memcpy(encm, bt_ciphertext + p_len, p_len);
+
+  ret = (int)out_len;
+end:
+  if (botan_pk_op_encrypt_destroy(op_ctx) ||
+      botan_pubkey_destroy(key) ||
+      botan_rng_destroy(rng)) {
+
+      // should never happen
+      (void)fprintf(stderr, "%s:%d ERROR when deinitializing\n", __FILE__, __LINE__);
+      ret = -1;
+  }
+
+  free(bt_ciphertext);
+  return ret;
 }
 
 int
 pgp_elgamal_private_decrypt(uint8_t *out,
-				const uint8_t *g_to_k,
+				const uint8_t *g2k,
 				const uint8_t *in,
 				size_t length,
 				const pgp_elgamal_seckey_t *seckey,
 				const pgp_elgamal_pubkey_t *pubkey)
 {
-	BIGNUM	*bndiv;
-	BIGNUM	*c1x;
-	BN_CTX	*tmp;
-	BIGNUM	*c1;
-	BIGNUM	*c2;
-	BIGNUM	*p;
-	BIGNUM	*x;
-	BIGNUM	*m;
-	int	 ret;
-
-	ret = 0;
-	/* c1 and c2 are in g_to_k and in, respectively*/
-	c1 = BN_bin2bn(g_to_k, (int)length, NULL);
-	c2 = BN_bin2bn(in, (int)length, NULL);
-	/* other bits */
-	p = pubkey->p;
-	x = seckey->x;
-	c1x = BN_new();
-	bndiv = BN_new();
-	m = BN_new();
-	tmp = BN_CTX_new();
-	if (!c1 || !c2 || !p || !x || !c1x || !bndiv || !m || !tmp) {
-		goto done;
-	}
-	/*
-	 * m = c2 / (c1^x)
-	 */
-	if (!BN_mod_exp(c1x, c1, x, p, tmp)) {
-		goto done;
-	}
-	if (!BN_mod_inverse(bndiv, c1x, p, tmp)) {
-		goto done;
-	}
-	if (!BN_mod_mul(m, c2, bndiv, p, tmp)) {
-		goto done;
-	}
-	/* result */
-	ret = BN_bn2bin(m, out);
-done:
-	if (tmp) {
-		BN_CTX_free(tmp);
-	}
-	if (m) {
-		BN_clear_free(m);
-	}
-	if (bndiv) {
-		BN_clear_free(bndiv);
-	}
-	if (c1x) {
-		BN_clear_free(c1x);
-	}
-	if (c1) {
-		BN_clear_free(c1);
-	}
-	if (c2) {
-		BN_clear_free(c2);
-	}
-	return ret;
+  botan_rng_t            rng          = NULL;
+  botan_privkey_t        key          = NULL;
+  botan_pk_op_decrypt_t  op_ctx       = NULL;
+  int                    ret          = -1;
+  size_t                 out_len      = 0;
+  size_t                 p_len        = 0;
+  uint8_t*               bt_plaintext = NULL;
+
+  if (botan_rng_init(&rng, NULL)) {
+
+      FAIL("Random initialization failure");
+  }
+
+  // Output len is twice an order of underlying group
+  if (botan_mp_num_bytes(pubkey->p->mp, &p_len) ||
+      (length != p_len)) {
+
+      FAIL("Wrong public key");
+  }
+
+  /* Max size of an output len is twice an order of underlying group (twice byte-size of p)
+   * Allocate all buffers needed for encryption and post encryption processing */
+  out_len = p_len*2;
+
+  bt_plaintext = calloc(out_len, 1);
+  if (!bt_plaintext) {
+
+      FAIL("Memory allocation failure");
+  }
+
+  if (botan_privkey_load_elgamal(&key, pubkey->p->mp, pubkey->g->mp, seckey->x->mp) ||
```

Comment:
I have similar concerns over `botan_privkey_load_elgamal` but I can't find the documentation for this one. Again, correct me if I'm wrong.